app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/navigationui/NavigationLauncherActivity.java

```diff
@@ -96,6 +107,32 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
     mapView.onCreate(savedInstanceState);
     mapView.getMapAsync(this);
     localeUtils = new LocaleUtils();
+    ReplayRouteDto routeFromJson = new GsonBuilder().create().fromJson(obtainJson("reroute.json"),
+      ReplayRouteDto.class);
+    ReplayLocationMapper mapper = new ReplayLocationMapper(routeFromJson.getLocations());
+    replayLocationDispatcher = new ReplayLocationDispatcher(mapper.toLocations());
```

Comment:
Going from `ReplayRouteDto` > `ReplayLocationMapper` > to finally `ReplayLocationDispatcher` feels like a good amount of setup.  Is there any way to streamline this setup?  

Is the point of having these separated to allow for different types of `Dto` objects?  Would adding a constructor that takes a `String` JSON make sense?  And then we would do the mapping behind the scenes.

Reply:
> Is the point of having these separated to allow for different types of `Dto` objects?

Yeah, kind of.

> Would adding a constructor that takes a `String` JSON make sense? And then we would do the mapping behind the scenes.

Nah, I don't think that would make sense because we'd be coupling the mapping within the `ReplayLocationDispatcher`.

The idea was having a class responsible of dispatching `Location`s (`ReplayLocationDispatcher`) agnostic to the data sources - that's why the replay location dispatcher only accepts a list of Android locations. Ultimately, we're interested in getting a list of locations and firing a callback wisely.

In the case of the replay JSON files we need `ReplayLocationDto` (so we can serialize/deserialize files with that concrete route JSON format) and then a mapper (`ReplayLocationMapper`) to convert "replay location" objects into Android `Location`s. IMO those classes need to be separated because they have clear and different responsibilities.

With this idea, we can take an entire route or portions/slices of routes and convert them into a list of Android locations and simulate them through the `ReplayLocationDispatcher`. Note that in these cases we don't need DTOs but we'd need classes that know how to generate "synthetic" `Location`s from the data they have (a.k.a. "mappers"). As you may know, we're already doing something similar in `MockLocationEngine` but it's coupled to `LocationEngine`. This is related to 👇 which I'd like to work on next.

> - [ ] Replace `MockLocationEngine` with `ReplayLocationEngine`?

That being said, I guess the overall confusion is more like a naming problem. What about renaming `ReplayRouteDto` and `ReplayLocationMapper` into `ReplayJsonRouteDto` and `ReplayJsonRouteLocationMapper` respectively? Would that make things clearer? What do you think?

Reply:
> That being said, I guess the overall confusion is more like a naming problem. What about renaming ReplayRouteDto and ReplayLocationMapper into ReplayJsonRouteDto and ReplayJsonRouteLocationMapper respectively? Would that make things clearer? What do you think?

@Guardiola31337 yeah thanks for the context here.  ^ I think the above makes sense as well.